handler/src/test/java/io/netty/handler/flow/FlowControlHandlerTest.java

```diff
@@ -0,0 +1,382 @@
+        GROUP.shutdownGracefully();
+    }
+
+    /**
+     * The {@link ByteToStringDecoder} decodes this {@code byte[]} into three messages.
+     */
+    private static ByteBuf newMessage() {
+        return Unpooled.wrappedBuffer(new byte[]{ (byte) '1', (byte) '2', (byte) '3' });
+    }
+
+    private Channel newServer(final boolean autoRead, final ChannelHandler... handlers) {
+        assertTrue(handlers.length >= 1);
+
+        ServerBootstrap serverBootstrap = new ServerBootstrap();
+        serverBootstrap.group(GROUP)
+            .channel(NioServerSocketChannel.class)
+            .option(ChannelOption.SO_REUSEADDR, true)
+            .childOption(ChannelOption.AUTO_READ, autoRead)
+            .childHandler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    ChannelPipeline pipeline = ch.pipeline();
+                    pipeline.addLast(new ByteToStringDecoder());
+                    pipeline.addLast(handlers);
+                }
+            });
+
+        return serverBootstrap.bind(0)
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    private Channel newClient(Channel server) {
+        Bootstrap bootstrap = new Bootstrap();
+
+        bootstrap.group(GROUP)
+            .channel(NioSocketChannel.class)
+            .option(ChannelOption.SO_REUSEADDR, true)
+            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 1000)
+            .handler(new ChannelInboundHandlerAdapter() {
+                @Override
+                public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                    ReferenceCountUtil.release(msg);
+                    throw new IllegalStateException("The client will never receive a response!");
+                }
+            });
+
+        return bootstrap.connect(server.localAddress())
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    /**
+     * This test demonstrates the default behavior if auto reading
+     * is turned on from the get-go and you're trying to turn it off
+     * once you've received your first message.
+     *
+     * NOTE: This test waits for the client to disconnect which is
+     * interpreted as the signal that all {@code byte}s have been
+     * transferred to the server.
+     */
+    @Test
+    public void testAutoReadingOn() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(3);
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                ReferenceCountUtil.release(msg);
+                latch.countDown();
+
+                // We're turning off auto reading in the hope that no
+                // new messages are being sent but that is not true.
+                ctx.channel().config().setAutoRead(false);
```

Comment:
What if on the first socket read only 1 byte is read (or less than 3 in this case)? It seems like you will never read all the data in this case.